### PR TITLE
Count mcsolve's no jump trajectory as an extra.

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
`mcsolve` runs with `improved_sampling` has a runs with no jump, therefore fully deterministic. This run is no longer counted in `ntraj`. 
This clear confusion when merging results with `improved_sampling`: 
```
res1 = mcsolver.run(psi0, ..., ntraj=100)
res2 = mcsolver.run(psi0, ..., ntraj=10)
result = res1 + res2
```
In this case, since the no jump trajectory is the same in both case, there is only 109 different trajectories. So if we merge keeping all random trajectories with the same weight, we get a `99:9` ratio instead of the `100:10` that we would naturally expect. To keep the intuitive ratio, we need to add relative weight between trajectories of each run.

By not counting the no-jumb trajectory in the provided `ntraj`, then expected ratio comes naturally.

It also help with reproducing result using seeds. To reproduce `result` from it's seeds, we would need to remove `seeds[100]` and run 109 trajectories. Keeping the no-jump appart, it does not use a seed and `result.seeds` can be used as is to reproduce the simulation.

I created `MultitrajResult.add_deterministic` for no jump trajectories. It does not take the seed, but take the absolute weight. And added the weight to the `add` method. By having the weight be part of the multi trajectory object instead of each trajectories, merging `MultitrajResult` is simpler since they can reuse the same single trajectory result.

### CODE REVIEW

**Comment by pmenczel** [comment]
Thank you Eric, this must have been a lot of work. I haven't had time yet to look at it in full detail, but I have a few preliminary comments / questions:
* Having the no-jump trajectory not count towards the `ntraj` seems like a good idea to me. In particular the handling of the seeds becomes much more sane.
* Also the change of having the weights part of the multi-result instead of the the trajectory-results makes a lot of sense in hindsight.
* What is the reason that the time-dependent weights are still part of the trajectory result? That seems a bit inconsistent to me and I wonder if the `TrajectoryResult` class could perhaps just be removed?
* We now have three names for more-or-less the same concept, "no jump", "deterministic" and "absolute". I guess the "no jump" name is used within the mcsolve module, which is fine. In the multitrajresult module, perhaps we could use either "deterministic" or "absolute" always? (Out of those, I would probably prefer "deterministic")

**Comment by Ericgig** [comment]
> * What is the reason that the time-dependent weights are still part of the trajectory result? That seems a bit inconsistent to me and I wonder if the `TrajectoryResult` class could perhaps just be removed?

The time-dependent weights are available as the `trace` attribute for each trajectories, so I though them as an (indirectly?) physical property of each trajectories.  Since they are per trajectories, if feels normal to belong to the single trajectory result. But moving the logic to `NmmcResult` would simplify `MultiTrajResult` a lot and allow to remove `TrajectoryResult`.

I will look into it this week.

**Comment by pmenczel** [comment]
> The time-dependent weights are available as the `trace` attribute for each trajectories, so I though them as an (indirectly?) physical property of each trajectories. Since they are per trajectories, if feels normal to belong to the single trajectory result. But moving the logic to `NmmcResult` would simplify `MultiTrajResult` a lot and allow to remove `TrajectoryResult`.

I do think it would be good to keep the `trace` property on the trajectories, but for the purpose of calculating the weighted averages, it could perhaps be treated like the other types of weights that we have.

**Review by pmenczel — COMMENTED** [review]
Thank you, it looks very good.

I noticed that the `MultiTrajResult` docstring might need some updates as well:
* `average_e_data` appears twice, but the second one should be `std_e_data`
* `final_state` should be `average_final_state` in the list of attributes
* The attributes `states`, `expect`, `e_data` and (after fixing the previous point) `final_state` are all missing, but that might be intentional?
* Other attributes not listed in the docstring: `fixed_weights`, `trajectories`, `deterministic_trajectories`, `num_trajectories`, `seeds`, `e_ops`

(I could write those, but I am not sure which of them, perhaps all, are left out intentionally.)

File: qutip/solver/mcsolve.py [hunk 1]
```diff
@@ -791,7 +765,7 @@ def _run_improved_sampling_mixed(
         result.initial_states = [self._restore_state(state, copy=False)
                                  for state, _ in ics_info.state_list]
         # add back +1 for the no-jump trajectories:
-        result.ntraj_per_initial_state = [(n+1) for n in ics_info.ntraj]
+        result.ntraj_per_initial_state = ics_info.ntraj
```
[the remark this answers is not in the document]
Comment (pmenczel) [reply qutip/solver/mcsolve.py:768]: Sorry the formatting is weird. I only meant to remove the comment.

File: qutip/solver/multitrajresult.py [hunk 2]
```diff
@@ -943,7 +918,49 @@ def merge(sum1, weight1, sum2, weight2):
         return new
 
 
-class McResult(MultiTrajResult):
+class _McBaseResult(MultiTrajResult):
+    # Collapse are only produced by mcsolve.
+    def _add_collapse(self, trajectory, *, rel=None, abs=None):
+        if rel is not None:
+            self.collapse.append(trajectory.collapse)
```
[the remark this answers is not in the document]
Comment (Ericgig) [reply qutip/solver/multitrajresult.py:937]: Deterministic trajectories are no-jump trajectories, so no collapse.
`runs_states` also only has non-deterministic runs data.

**Comment by Ericgig** [comment]
I added more entries in the docstring.
Since there was no way to get the deterministic trajectory's states and expect, I made it so that these trajectories are always kept.

File: qutip/solver/mcsolve.py [hunk 3]
```diff
@@ -511,13 +511,11 @@ def _initialize_stats(self):
         })
         return stats
 
-    def _no_jump_simulation(self, state, tlist, e_ops, seed=None):
+    def _no_jump_simulation(self, state, tlist, e_ops):
         """
         Simulates the no-jump trajectory from the initial state `state0`.
         Returns a tuple containing the seed, the `Result` describing
         this trajectory, and the trajectory's probability.
```
Comment (pmenczel) [inline qutip/solver/mcsolve.py:518]:
```suggestion
        Returns a tuple containing the `Result` describing
        this trajectory and the trajectory's probability.
```

File: qutip/solver/multitrajresult.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmenczel) [reply qutip/solver/multitrajresult.py:954]: Oh, you're right, I thought `runs_states` had the deterministic ones as well

[the remark this answers is not in the document]
Comment (pmenczel) [reply qutip/solver/multitrajresult.py:954]: Oh you're right, I thought `runs_states` had the deterministic ones as well.

File: qutip/solver/mcsolve.py (hunk 3 above)
Comment (pmenczel) [inline qutip/solver/mcsolve.py:518]:
```suggestion
        Returns a tuple containing the `Result` describing
        this trajectory and the trajectory's probability.
```

**Review by pmenczel — APPROVED** [review]
Thank you. I think the `_merge_weight` function in MultiTrajResult could be removed now. Other than that, approved by me.

**Pull request merged**